Trinity long run Job

Bhargavasomu opened the pull request:
### What was wrong?
Fixes #170 


### How was it fixed?
Now a new Circle CI job is created to run `trinity --ropsten` for 90 seconds.

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://www.millionfeed.com/wp-content/uploads/2017/09/incredibly-cute-baby-animal-pictures-around-the-world-15065608488gn4k.jpg)

cburgdorf commented on tests/trinity_long_run/test_trinity_long_run.py:
```diff
@@ -0,0 +1,48 @@
+
+# This fixture provides a tear down to run after each test that uses it.
+# This ensures the AsyncProcessRunner will never leave a process behind
+@pytest.fixture(scope="function")
```
We already have this fixture, if we move it to `conftest`, we can reuse it in multiple test files that need it.

cburgdorf commented on tests/trinity_long_run/test_trinity_long_run.py:
```diff
@@ -0,0 +1,48 @@
+    )
+)
+@pytest.mark.asyncio
+async def test_does_not_throw_long_run(async_process_runner, command):
```
I think this just copies the existing `test_does_not_throw` test. Why would we not just increase the timeout on the existing one?

cburgdorf replied:
My question is, why would we need a new job at all if it is exactly the same just with a higher timeout? Why not just increase the timeout of the existing one?

>does parallelization among jobs better than parallellization in a particular Job
Also, if you are referring to frequent CI fails with these tests, then there's a high chance they will run much smoother if rebased against latest master (we just merged a bugfix that was causing frequent CI fails with these tests)

Bhargavasomu replied:
@cburgdorf the reasoning that I had was as follows. If we just increase the timeout for the existing one, then all of them will happen serially, instead of parallel. Hence I thought that maybe we could have a new job because `Circle CI` runs parallel jobs well. Please correct if this approach is deluded.

cburgdorf replied:
We could put the `timeout` into the test parameters and then change them to something like this.

```
    (
        # mainnet
        (20, 'trinity',),
        ('20, trinity', '--tx-pool',),
        ('20, trinity', '--sync-mode=light',),
        # ropsten
        # we run the first one for 90 seconds to catch potential bugs that may only occur after a while
        ('90, trinity', '--ropsten',),
        (20, 'trinity', '--ropsten', '--tx-pool',),
        (20, 'trinity', '--sync-mode=light', '--ropsten',),
    )
```

Then, all tests would run as they run today except that one of them would run slightly longer. It does make the job run *70* seconds longer but I'm not sure if this really needs it's own job for better concurrency.

If you do feel strongly about making this an extra job for better concurrency then at least, we should refactor the code so that we move the body of the test to some function that we can reuse to not copy for the new job.

veox replied:
> It does make the job run 70 seconds longer (...)

It makes the parametrised test run 70 seconds longer in every CI job that selects the test. So,  py3{6,7}-{integration,wheel-cli} - 4 CI jobs; 280 seconds higher total run time.

cburgdorf replied:
>4 CI jobs; 280 seconds higher total run time.

Right. With 4 jobs, that does sound like something we should optimize for. Then, I'd say, let's refactor the test code to be reusable for another job. And maybe even just stick to running against Python 3.6 for this special job to keep things light.

Bhargavasomu commented:
@cburgdorf done, please review. Thankyou.

veox commented:
"How was it fixed?" section is now outdated.

Bhargavasomu commented:
@cburgdorf is there anyway where we could apply `pytest parametrizations` on functions which are imported but not explicitly defined at the place where we are applying `pytest parametrization`. (I searched quite a bit, but couldn't find anything). Something on the lines of 
```
from x import func

@pytest.mark.parametrize(
    ...
)
func
```

If we can't do it in the above way, then refactoring the code to a new function and using it won't be of any use.

veox commented:
It is tests that can be parametrised, not any function.

What is needed here is two `test_does_not_throw[,_long_run]()` functions (tests), as you had them initially, but with the body of the test factored out (in the same file) - e.g. `_run_does_not_throw()`.

For existing CI jobs, `test_does_not_throw_long_run` can be excluded in CI configuration explicitly. A new CI job needs to be added to run this test only.

Bhargavasomu commented:
@cburgdorf could you do a preliminary review? If everything is fine, then I will create a `Circle CI` job.

cburgdorf commented on tests/conftest.py:
```diff
@@ -118,6 +121,22 @@ def event_loop():
+    )
+    yield runner
+    try:
+        if hasattr(runner, 'proc'):
```
This stands out to me for two reasons:

1. If we need this, I think it should be handled directly within the `def kill` of `AsyncProcessRunner`.

2. Do you know in which circumstances this happens? Because, this can only mean we are trying to call `kill` on an `AsyncProcessRunner` that we did not call `run` on before which makes me think we are sweeping a problem under the rug here

Bhargavasomu replied:
@cburgdorf the reason why I had to put this here, is because of the following scenario that I was seeing. When I moved `async_process_runner` to `conftest.py`, I was seeing this error that the object doesn't have the attribute `proc`, which means that it is trying to kill a process which it even hasn't run. I believe it is some issue that deals with `pytest` and `conftest.py`, but I don't know how they really work. What should we do here? Instead of moving `async_process_runner` to `conftest.py`, should we just duplicate the fixture in both test files?

cburgdorf replied:
>should we just duplicate the fixture in both test files?

If we can not figure out the reason for this behavior then, yes, I'd rather duplicate that fixture than sweeping things under the rug here.

cburgdorf commented on tests/integration/helpers.py:
```diff
@@ -0,0 +1,20 @@
+async def _run_trinity_for_errors(async_process_runner, command, time):
+    # This is our last line of defence. This test basically observes the first `time` seconds of the
```
1. I think the comment should go back at the actual test

2. This function can have a docstring such as 

```
"""
Run the given ``command`` on the given ``async_process_runner`` for ``time`` seconds and throw an Exception in case any unresolved Exceptions are detected in the output of the command.
"""
```

3. Let's rename this to `run_command_and_detect_errors`

Bhargavasomu commented:
@cburgdorf me forgetting the `await` keyword before the `run_command_and_detect_errors` caused the lot of errors. I have fixed it and moved the `async_process_runner` to `conftest.py`. I am pretty confident that this is correct and hence moving on with the creation of another `Circle CI` job.

Bhargavasomu commented:
@cburgdorf I have created the `Circle CI` job as well. Please review it now. Thankyou.

cburgdorf commented on tox.ini:
```diff
@@ -97,6 +98,12 @@ deps = {[common-integration]deps}
+deps = {[common-integration]deps}
+passenv = {[common-integration]passenv}
+commands =
+    pytest --integration -n 1 {posargs:tests/trinity_long_run/ -k 'not lightchain_integration'}
```
Since this is served from its own directory `trinity_long_run` (which makes sense to not bloat excludes further), you can drop the `-k 'not lightchain_integration'` here. And I think you can drop the `--integration` here as well.

Bhargavasomu commented:
@cburgdorf done. Please take a look now. Thankyou.

cburgdorf commented:
Thanks! Merged :tada: :clap:
